usecases/schema/cache.go

```diff
@@ -163,3 +170,103 @@ func (s *schemaCache) deleteClassState(name string) {
 	defer s.Unlock()
 	delete(s.ShardingState, name)
 }
+
+func (s *schemaCache) unsafeFindClassIf(pred func(*models.Class) bool) *models.Class {
+	for _, cls := range s.ObjectSchema.Classes {
+		if pred(cls) {
+			return cls
+		}
+	}
+	return nil
+}
+
+func (s *schemaCache) unsafeFindClass(className string) *models.Class {
+	for _, c := range s.ObjectSchema.Classes {
+		if c.Class == className {
+			return c
+		}
+	}
+	return nil
+}
+
+func (s *schemaCache) addClass(c *models.Class, ss *sharding.State) {
+	s.Lock()
+	defer s.Unlock()
+
+	s.ShardingState[c.Class] = ss
+	s.ObjectSchema.Classes = append(s.ObjectSchema.Classes, c)
+}
+
+func (s *schemaCache) updateClass(u *models.Class, ss *sharding.State) error {
+	s.Lock()
+	defer s.Unlock()
+
+	if c := s.unsafeFindClass(u.Class); c != nil {
+		*c = *u
+	} else {
+		return errClassNotFound
+	}
+	if ss != nil {
+		s.ShardingState[u.Class] = ss
+	}
+	return nil
+}
+
+func (s *schemaCache) addProperty(class string, p *models.Property) (models.Class, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	c := s.unsafeFindClass(class)
+	if c == nil {
+		return models.Class{}, errClassNotFound
+	}
+
+	// update all at once to prevent race condition with concurrent readers
+	src := c.Properties
+	dest := make([]*models.Property, len(src)+1)
+	copy(dest, src)
+	dest[len(src)] = p
+	c.Properties = dest
+	return *c, nil
+}
+
+// readOnlySchema returns a read only schema
+// Changing the schema outside this package might lead to undefined behavior.
+func (s *schemaCache) readOnlySchema() *models.Schema {
+	s.RLock()
+	defer s.RUnlock()
+	return shallowCopySchema(s.ObjectSchema)
+}
+
+func (s *schemaCache) readOnlyClass(name string) (*models.Class, error) {
+	s.RLock()
+	defer s.RUnlock()
+	c := s.unsafeFindClass(name)
+	if c == nil {
+		return nil, errClassNotFound
+	}
+	cp := *c
```

Comment:
I'm not sure if this is enough. AFAIK this only copied the primitive values but still references the structs/slices etc.

I added in the last try a deepcopy package here: entities/deepcopy/models_deepcopy.go and it worked back then, but I haven't looked at it sinc

Reply:
As the name of the function `readOnlyClass()` suggests, this is a read-only copy. The only attribute that can potentially cause a race condition is 'class.properties.' This particular slice is not modified during the schema update process because a new one is created with each update